server/cron/cron.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2022 Woodpecker Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cron
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/robfig/cron"
+	"github.com/rs/zerolog/log"
+
+	"github.com/woodpecker-ci/woodpecker/server"
+	"github.com/woodpecker-ci/woodpecker/server/model"
+	"github.com/woodpecker-ci/woodpecker/server/pipeline"
+	"github.com/woodpecker-ci/woodpecker/server/store"
+)
+
+const (
+	// checkTime specifies the interval woodpecker checks for new cron jobs to exec
+	checkTime = 10 * time.Second
+
+	// checkItems specifies the batch size of jobs to retrieve per check from database
+	checkItems = 10
+)
+
+// Start starts the cron scheduler loop
+func Start(ctx context.Context, store store.Store) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(checkTime):
+			go func() {
+				now := time.Now()
+				log.Trace().Msg("Cron: fetch next jobs")
+
+				jobs, err := store.CronListNextExecute(now.Unix(), checkItems)
+				if err != nil {
+					log.Error().Err(err).Int64("now", now.Unix()).Msg("obtain cron job list")
+					return
+				}
+
+				for _, job := range jobs {
+					if err := runJob(job, store, now); err != nil {
+						log.Error().Err(err).Int64("jobID", job.ID).Msg("run cron job failed")
+					}
+				}
+			}()
+		}
+	}
+}
+
+// CalcNewNext parses a cron string and calculates the next exec time based on it
+func CalcNewNext(schedule string, now time.Time) (time.Time, error) {
+	// remove local timezone
+	now = now.UTC()
+
+	// TODO: let timezone set as server setting
+	//       will UI adjustments and server flag
```

Comment:
So we use UTC now as suggested. What do you mean by this TODO? Allowing the server admin to set a specific timezone?

Reply:
yes but that need more preparation ... that's an extention for another pull

Reply:
```suggestion
	// TODO: allow the users / the admin to set a specific timezone
```